File: c++/classes/inheritance/multilevel_inheritance.cpp
#include <iostream>

using std::string, std::cout, std::endl, std::boolalpha, std::noboolalpha;

class Mammal {
public:
  bool is_domesticated;
  bool has_tail;

  static bool is_milk_feeded;
  static string class_taxon;
};

bool Mammal::is_milk_feeded = true;
string Mammal::class_taxon = "Mammalia";

class Fox : public Mammal {
public:
  string name;

  static bool has_fur;
  static string species_taxon;

  Fox() {
    has_tail = true;
    is_domesticated = false;
  }

  void show() {
    cout << "Hi, my name is " << name << "." << endl
      << "Class: " << class_taxon << endl
      << "Species: " << species_taxon << endl
      << boolalpha
      << "Am I milk-feeded as a cub? " << is_milk_feeded << endl
      << "Do I have tail? " << has_tail << endl
      << "Do I have fur? " << has_fur << endl
      << "Am I domesticated? " << is_domesticated << endl
      << noboolalpha;
  }
};

bool Fox::has_fur = true;
string Fox::species_taxon = "Vulpes vulpes";

class PetFox : public Fox {
public:
  PetFox() {
    is_domesticated = true;
  }
};

int main() {
  Fox fox_ferdinand;
  fox_ferdinand.name = "Ferdinand";

  fox_ferdinand.show();
  cout << endl;

  PetFox fox_agnes;
  fox_agnes.name = "Agnes";

  fox_agnes.show();
  cout << endl;
}

File: pascal/classes/inheritance/multilevel_inheritance.pas
{$MODE OBJFPC}

program MultilevelInheritance;

type
  Mammal = class
  public
    is_domesticated: boolean;
    has_tail: boolean;

    is_milk_feeded: boolean; static;
    class_taxon: string; static;

    constructor create();
  end;

  Fox = class(Mammal)
  public
    name: string;

    has_fur: boolean; static;
    species_taxon: string; static;

    constructor create();
    procedure show();
  end;

  PetFox = class(Fox)
  public
    constructor create();
  end;

constructor Mammal.create();
begin
  is_milk_feeded := true;
  class_taxon := 'Mammalia';
end;

constructor Fox.create();
begin
  Inherited Create;

  is_domesticated := false;
  has_tail := true;
  has_fur := true;
  species_taxon := 'Vulpes vulpes';
end;

constructor PetFox.create();
begin
  Inherited Create;
  
  is_domesticated := true;
end;

procedure Fox.show();
begin
  writeln('Hi, my name is ', name, '.');
  writeln('Class: ', class_taxon);
  writeln('Species: ', species_taxon);
  writeln('Am I milk-feeded as a cub? ', is_milk_feeded);
  writeln('Do I have tail? ', has_tail);
  writeln('Do I have fur? ', has_fur);
  writeln('Am I domesticated? ', is_domesticated);
end;

var
  fox_ferdinand: Fox;
  fox_agnes: PetFox;

begin
  fox_ferdinand := Fox.create();
  fox_ferdinand.name := 'Ferdinand';

  fox_ferdinand.show();
  writeln();

  fox_agnes := PetFox.create();
  fox_agnes.name := 'Agnes';

  fox_agnes.show();
  writeln();
end.
